source/common/router/header_parser.cc

```diff
@@ -1,28 +1,206 @@
 #include "common/router/header_parser.h"
 
+#include <ctype.h>
+
+#include <string>
+
+#include "common/common/assert.h"
 #include "common/protobuf/utility.h"
 
+#include "absl/strings/str_join.h"
+#include "absl/strings/str_split.h"
+
 namespace Envoy {
 namespace Router {
 
 namespace {
 
+enum class ParserState {
+  Literal,                   // processing literal data
+  VariableName,              // consuming a %VAR% name
+  ExpectArray,               // expect starting [ in %VAR([...])%
+  ExpectString,              // expect starting " in array of strings
+  String,                    // consuming an array element string
+  ExpectArrayDelimiterOrEnd, // expect array delimiter (,) or end of array (])
+  ExpectArgsEnd,             // expect closing ) in %VAR(...)%
+  ExpectVariableEnd          // expect closing % in %VAR(...)%
+};
+
+std::string unescape(absl::string_view sv) { return absl::StrJoin(absl::StrSplit(sv, "%%"), "%"); }
+
+// Implements a state machine to parse custom headers. Each character of the custom header format
```

Comment:
very nice and readable code in general.  I like this better than the regex.  Also I think it can potentially be made more general, as described below.

However I think this parser does deserve some rigorous unit tests :)

Reply:
I added enough tests to exercise each case in the switch statement and the branches within them. Is there a particular part you'd like more test cases for?